### PR TITLE
NonStop port updates for 3.0.0.

### LINKED ISSUE #5087: Assistance request: setting up for NonStop platform support
Greetings! I'm about to submit a PR for the NonStop NSX and NSE platform ports for OpenSSL and need assistance in setting up required pieces (Configurations/x-nonstop.conf) as well as documenting the port. The overall port has two basic aspects that deviate from standard code:

- Inclusion of platform FLOSS method redirects to emulate Linux behaviour, and
- Changes to root id - root on NonStop is 65535 rather than 0, and
- Compiles under c99 rather than gcc.

### PR DESCRIPTION
**Pull request by rsbeckerca**
HPE NonStop Port Changes for 3.0.0  Includes unthreaded, PUT, and SPT for OSS.

The port changes include wrapping where necessary for FLOSS and
appropriate configuration changes to support that. Two tests
are excluded as being inappropriate for the platform.

The changes are:
* Added /usr/local/include to nonstop-nsx_spt_floss to load floss.h
* Added SPT Floss variant for NonStop
* Wrapped FLOSS definitions in OPENSSL_TANDEM_FLOSS to allow selective enablement.
* SPT build configuration for NonStop
* Skip tests not relevant for NonStop
* PUT configuration changes required for NonStop platforms
* Configurations/50-nonstop.conf: updates for TNS/X platform.
* FLOSS instrumentation for HPE NonStop TNS/X and TNS/E platforms.
* Configurations/50-nonstop.conf: modifications for non-PUT TNS/E platform b
* Fix use of DELAY in ssltestlib.c for HPNS.
* Fixed commit merge issues and added floss to http_server.c

CLA: Permission is granted by the author to the OpenSSL team to use these modifications.
Fixes #5087.

Signed-off-by: Randall S. Becker <rsbecker@nexbridge.com>
##### Checklist
- [X] documentation is added or updated
- [X] tests are added or updated

### CODE REVIEW

**Comment by rsbeckerca** [comment]
CLA should be on file - it was submitted months ago. This PR can be kept open until needed/approved. It is for reference for the NonStop community.

**Comment by mattcaswell** [comment]
CLA issue should be fixed now. Close/reopen to kick CLA bot.

File: test/bntest.c
Comment (slontis) [inline test/bntest.c:14]: What is this include required for? strcasecmp?

Comment (rsbeckerca) [reply test/bntest.c:14]: Correct. For some historical reason, `strcasecmp` and related string functions of this form are in `strings.h` on this platform.

Comment (slontis) [reply test/bntest.c:14]: Could you possibly put the dependency as a comment for these e.g
#include <strings.h> /* strcasecmp */
strcasecmp needs a cleanup at some point..
Some files are including "e_os.h", whilst others are doing this horrible #ifdef for windoze.

Comment (slontis) [reply test/bntest.c:14]: Also the notation in openssl is to indent
e.g:
```
#ifdef X
# include Y
# ifdef
#  include
# endif
```

File: include/openssl/e_os2.h
Comment (slontis) [inline include/openssl/e_os2.h:144]: This is a public header.. Are you sure you want it here?

File: test/bntest.c
Comment (rsbeckerca) [reply test/bntest.c:14]: Sure. Will fix these and others in the PR.

File: include/openssl/e_os2.h
Comment (rsbeckerca) [reply include/openssl/e_os2.h:144]: The `nssgetpid()` method is a redirect of `getpid()` across the board. The platform port actually runs in a non-POSIX environment as well as the POSIX flavour. This particular code has been around a while in the port. I would rather this was elsewhere too, but `getpid()`, in fact `pid_t` also, is not available for that variant. The use of `getpid()` is rather pervasive so I think the choice was to do it here rather than everywhere to minimize the code impact. Is there a better place for this and the follow-on definitions?

**Comment by rsbeckerca** [comment]
Port is updated based on comments above.

File: Configurations/50-nonstop.conf
Comment (levitte) [inline Configurations/50-nonstop.conf:19]: How about `add(qw(-Wverbose -I/usr/local/include))`?

The same question goes for all places where you've added this kind of sub...

File: Configurations/50-nonstop.conf
Comment (levitte) [inline Configurations/50-nonstop.conf:20]: `shared_cflag` are additional flags for shared libraries and modules.  You shouldn't need to copy the `cflags` value here.

File: Configurations/50-nonstop.conf
Comment (levitte) [inline Configurations/50-nonstop.conf:22]: Just like `shared_cflag` is a complement for `cflags`, `shared_ldflag` is a complement for `lflags`, i.e. it shouldn't be necessary to repeat...  however, seeing the additions in the latter, I gotta ask if `-Wxld` is a cumulative flag, or if it's rather a "last man standing" type of flag?  The latter could prove problematic.

Furthermore, if `-export_all` does what I think it does, I gotta ask, is there no way to select which symbols are going to be exported?

File: Configurations/50-nonstop.conf
Comment (levitte) [inline Configurations/50-nonstop.conf:15]: If you add the following here, the macro `OPENSSL_SYS_TANDEM` will automatically be created for you:

``` perl
        sys_id           => 'TANDEM',
```

Also, every `OPENSSL_SYSNAME_TANDEM` in the C source should be renamed to `OPENSSL_SYS_TANDEM`, and you can remove `"OPENSSL_SYSNAME_TANDEM"` from every `defines` in this file.

Furthermoore, it might be a good idea to collect all the common `defines` values in one here, and please use `add()`, like this (here):

``` perl
        defines    => add("OPENSSL_VPROC=\$(OPENSSL_VPROC)", "_XOPEN_SOURCE", "_XOPEN_SOURCE_EXTENDED=1", "B_ENDIAN", "OPENSSL_SYSNAME_TANDEM"),
```

File: include/openssl/e_os2.h
Comment (levitte) [reply include/openssl/e_os2.h:145]: The point is that you have the exact same code in `e_os.h` and `e_os2.h`, which seems unnecessary.

If there are things that are for internal OpenSSL use, then that should be in `e_os.h` *only*.
If there are things meant to be for public consumption, i.e. something affect our public headers, then that should be in `e_os2.h` *only*
(if some goes in one and some in the other, you'll get to split some of the code apart)

**Comment by levitte** [comment]
Other than my comments, I don't have much to say.  This is much cleaner than I expected!

File: Configurations/50-nonstop.conf
Comment (rsbeckerca) [reply Configurations/50-nonstop.conf:19]: I'm not sure I really need the `-Wverbose` option. In fact, if the '-I/usr/local/include' is redundant, I could remove it all.

File: Configurations/50-nonstop.conf
Comment (rsbeckerca) [reply Configurations/50-nonstop.conf:22]: Will fix up the `shared_cflag` and `cflag` usage as well as `shared_ldflag` and `ldflag`.

The `-Wxld` is similar to the `-Wrld`. Loader option. I don't think it is positional. Will check.

The `-export_all` does what you think it does, but only exports globals. static variables and procedures are not exported. There is a keyword in the platform's c99 (`export$`) that could be used instead but it is on a per-symbol basis in C source. I have not successfully used that. Otherwise it is possible to selectively specify each symbol on the command line, or in a linker input file - they we get into the same problem as `ar` and `.so` construction limits.

File: include/openssl/e_os2.h
Comment (rsbeckerca) [reply include/openssl/e_os2.h:145]: I am pretty sure (maybe 60%-ish) that the `e_os2.h` code is referenced in a calling application wrapper in one situation - not part of the port but used by it. I don't have access to that application, so I'm reluctant to move the code out of public.

The config and related changes are going to take a while to get right and I'm not going to be able to get there until after Sept 16th (although I'll try).

File: Configurations/50-nonstop.conf
Comment (rsbeckerca) [reply Configurations/50-nonstop.conf:15]: This is a pretty big change. I like it a lot, however. It should not be a problem to do. Is there a deadline on this PR?

File: Configurations/50-nonstop.conf
Comment (levitte) [reply Configurations/50-nonstop.conf:22]: We do have support for linker input files that specify what symbols should be exported...

Comment (rsbeckerca) [reply Configurations/50-nonstop.conf:22]: The `-Wxld` is how parameters are passed to the linker through the c99 front end. It is possible to specify multiple copies of this AFAIK.

Comment (rsbeckerca) [reply Configurations/50-nonstop.conf:22]: The `-exported_symbol` option to the linker can be specified in a linker input file, it would look something like:

```
-exported_symbol A
-exported_symbol B
etc
```
Passed in through the `-Wxld_obey=file` option.

**Comment by rsbeckerca** [comment]
I might need a little help on why the Travis CI build failed. Something about the `nssgetpid` symbol isn't being liked on Mac/OS but it should not even be there for that build.

File: Configurations/50-nonstop.conf
Comment (levitte) [reply Configurations/50-nonstop.conf:22]: `-exported_symbol` looks useful together with `-Wxld_obey=file`. I can probably concoct something tomorrow that should help.

File: Configurations/50-nonstop.conf
Comment (levitte) [reply Configurations/50-nonstop.conf:15]: Hmmm, there is, actually, more or less tomorrow . I could simply approve as is, and we have another PR to streamline it.

File: include/openssl/e_os2.h
Comment (levitte) [reply include/openssl/e_os2.h:145]: Why should openssl carry stuff publically that concerns a completely different source set? That tells me that the application you're talking about should handle its own and that the code added in `e_os2.h` isn't needed for openssl needs, neither public nor private.

File: Configurations/50-nonstop.conf
Comment (kaduk) [reply Configurations/50-nonstop.conf:15]: My understanding is that in order to make 3.0 it needs to be in an approved state by today, basically.

File: include/openssl/e_os2.h
Comment (rsbeckerca) [reply include/openssl/e_os2.h:145]: > 
> 
> Why should openssl carry stuff publically that concerns a completely different source set? That tells me that the application you're talking about should handle its own and that the code added in `e_os2.h` isn't needed for openssl needs, neither public nor private.

Briefly - I could go into much more detail in private - there are two personalities on NonStop: GUARDIAN (legacy) and OSS (POSIX-like). I have been maintaining the POSIX-side port of OpenSSL since early 2015 (from 1.0.2-ish). The GUARDIAN side, where there are no pids - you would see `cpu` and `pin` in the `nssgetpid()` code - has a very different interface layer and wrapper to run there. The `nssgetpid()` call is a bridge to allow SSL processes within the GUARDIAN environment to be managed, so it needs to be externalized for that environment.

File: Configurations/50-nonstop.conf
Comment (rsbeckerca) [reply Configurations/50-nonstop.conf:15]: > 
> 
> Hmmm, there is, actually, more or less tomorrow . I could simply approve as is, and we have another PR to streamline it.

I will commit to that streamlining (pun intended).

**Comment by rsbeckerca** [comment]
> 
> 
> I might need a little help on why the Travis CI build failed. Something about the `nssgetpid` symbol isn't being liked on Mac/OS but it should not even be there for that build.

It looks like `nssgetpid()` is missing from `util/libcrypto.num`. Without understanding the meaning of this, and because this procedure is only for the NonStop port, I do not feel comfortable modifying the contents. May I request some help here?

File: e_os.h
Comment (levitte) [inline e_os.h:307]: This line isn't really needed as far as I can tell.

File: e_os.h
Comment (levitte) [inline e_os.h:312]: This should be `static inline`

File: e_os.h
Comment (levitte) [inline e_os.h:327]: These two lines should trade places

File: include/openssl/e_os2.h
Comment (levitte) [reply include/openssl/e_os2.h:145]: It's still just for OpenSSL, and as far as I can tell, this shouldn't be needed in `e_os2.h` at all, the stuff in `e_os.h` should be enough.

If there are OpenSSL files that don't compile because of a missing `getpid()`, they should include `"e_os.h"` to get the proper definition.

**Comment by levitte** [comment]
The main blocker for getting this approved is the added code in `include/openssl/e_os2.h`, which really seem to be internal only.  Please remove it.

If other (non-OpenSSL) applications need nssgetpid() as well, they should really do the same trick themselves...  or someone might come up with an emulation library that does this and that everyone can depend on.

**Comment by rsbeckerca** [comment]
Running build/test with the e_os2.h changes

File: e_os.h
Comment (rsbeckerca) [reply e_os.h:327]: > 
> 
> These two lines should trade places

I'm not sure about that. The `nssgetpid(a)` is defined outside the `NSSGETPID_MACRO` block. I'm actually not sure why the person who did this did that. But I'll go with your change until I get yelled at.

File: e_os.h
Comment (rsbeckerca) [reply e_os.h:312]: I will check with the team. Can we fix this later? The person is away.

Comment (levitte) [reply e_os.h:312]: We can fix that later

File: e_os.h
Comment (levitte) [reply e_os.h:327]: Er, no. It's defined inside that block, not outside. It may be declared outside of it, but that's unnecessary.

**Comment by rsbeckerca** [comment]
> 
> 
> Running build/test with the e_os2.h changes

Pushed what I could.

File: include/openssl/e_os2.h
Comment (rsbeckerca) [reply include/openssl/e_os2.h]: Should be done in commit  2716b338b9.

**Comment by levitte** [comment]
If Travis doesn't complain, I'll approve

**Comment by rsbeckerca** [comment]
> 
> 
> If Travis doesn't complain, I'll approve

I'm monitoring it now. Note that I'll might be AFK for a bit (unsuccessfully trying my best to be on vacation for a week, whatever that means in today's weirdness).

**Comment by levitte** [comment]
> (unsuccessfully trying my best to be on vacation for a week, whatever that means in today's weirdness).

Oh boy do I know the feeling

**Comment by rsbeckerca** [comment]
> 
> 
> > (unsuccessfully trying my best to be on vacation for a week, whatever that means in today's weirdness).
> 
> Oh boy do I know the feeling

At this rate, Travis won't be done until I'm back ;)

**Review by levitte — APPROVED** [review]
This is good for now

File: crypto/dso/dso_dlfcn.c
Comment (slontis) [inline crypto/dso/dso_dlfcn.c:35]: fix the indent on this one..

File: crypto/rand/rand_egd.c
Comment (slontis) [inline crypto/rand/rand_egd.c:57]: This needs to be nested

#-if

and then the defines within 
#--define

(where - is a space)

File: crypto/rand/rand_egd.c
Comment (slontis) [inline crypto/rand/rand_egd.c:75]: char *transport

File: crypto/rand/rand_egd.c
Comment (slontis) [inline crypto/rand/rand_egd.c:79]: 4 space indenting here..

File: crypto/rand/rand_egd.c
Comment (slontis) [inline crypto/rand/rand_egd.c:84]: transport != NULL

File: crypto/rand/rand_egd.c
Comment (slontis) [inline crypto/rand/rand_egd.c:86]: (family, type, protocol);

Comment (slontis) [reply crypto/rand/rand_egd.c:86]: There are quite a few more of these.

File: crypto/rand/rand_egd.c
Comment (slontis) [inline crypto/rand/rand_egd.c:115]: Is this a global??

File: crypto/rand/rand_egd.c
Comment (slontis) [inline crypto/rand/rand_egd.c:138]: , AF_

File: crypto/rand/rand_egd.c
Comment (slontis) [inline crypto/rand/rand_egd.c:171]: does this ever need to reset to zero?

File: e_os.h
Comment (slontis) [inline e_os.h:300]: Spacing look wrong on these
should be starting with 
# ifdef

File: include/internal/cryptlib.h
Comment (slontis) [inline include/internal/cryptlib.h:219]:
```
#      include
```

File: crypto/rand/rand_egd.c
Comment (rsbeckerca) [reply crypto/rand/rand_egd.c:171]: > 
> 
> does this ever need to reset to zero?

I don't think it is reset. It is a first attempt. I think it is used as a boolean but I have to get this answered by someone else on the team.

**Comment by rsbeckerca** [comment]
Review changes requested by @slontis have been pushed as commit 83de323fc5.

File: crypto/rand/rand_egd.c
Comment (slontis) [inline crypto/rand/rand_egd.c:85]: indent is still not right on the inner blocks :)

Comment (slontis) [reply crypto/rand/rand_egd.c:85]: Also for single block ifs remove the {}

File: e_os.h
Comment (slontis) [inline e_os.h:326]: one more space here..

**Review by slontis — COMMENTED** [review]
Apologies for the NITS..

File: crypto/rand/rand_egd.c
Comment (rsbeckerca) [reply crypto/rand/rand_egd.c:85]: Argh. 3am. Sorry.

**Comment by rsbeckerca** [comment]
> 
> 
> Apologies for the NITS..

NITS are fine. My inexperience (and being the "front person" for this).

File: crypto/rand/rand_egd.c
Comment (levitte) [reply crypto/rand/rand_egd.c]: Note: if this was a multi-statement block, the starting brace should be on the same line as the condition (strictly speaking, its end), some would have had to be this:

``` C
    if (!_arg_present(transport) || transport != NULL || transport[0] == '\0') {
        /* yada yada yada */
    }
```

Comment (rsbeckerca) [reply crypto/rand/rand_egd.c]: Understood. Same style used at my $DAYJOB.

**Comment by rsbeckerca** [comment]
I don't think the Travis build failure was a result of my changes. This looks like a VM update issue. Please advise. Note: I am having connectivity issues (in back woods on walkabout).

**Comment by levitte** [comment]
> I don't think the Travis build failure was a result of my changes.

Confirmed

**Review by levitte — APPROVED** [review]
Note: I intend to follow up in this with a PR that streamlines this contribution, with @rsbecker's assistance.
For the moment, the primary interest is to get this in at all before beta 1 is released.

File: crypto/rand/rand_egd.c
Comment (mattcaswell) [inline crypto/rand/rand_egd.c:167]: Nit: No parens for return, i.e. just `return -1`.

**Review by mattcaswell — CHANGES_REQUESTED** [review]
I spotted a minor nit.

From some of the comments you made above, am I correct that some of these commits were actually authored by other people? Have *all* of those people signed CLAs?

File: crypto/rand/rand_egd.c
Comment (richsalz) [reply crypto/rand/rand_egd.c:167]: and the return goes on its own line, n'est ce pas ?

Comment (rsbeckerca) [reply crypto/rand/rand_egd.c]: Nit fixed.

The original effort to port 1.1.1 was done (and shared) as part of #5087 for 1.1.1. The 3.0.0 port is 100% my effort alone.

**Comment by rsbeckerca** [comment]
> 
> 
> Note: I intend to follow up in this with a PR that streamlines this contribution, with @rsbecker's assistance.
> For the moment, the primary interest is to get this in at all before beta 1 is released.

Would you like me to open the "Streamline NonStop Configuration" PR as a defect/enhancement? It's certainly not a nit.

**Comment by levitte** [comment]
> Would you like me to open the "Streamline NonStop Configuration" PR as a defect/enhancement? It's certainly not a nit.

Enhancement.

**Review by slontis — APPROVED** [review]

**Comment by rsbeckerca** [comment]
> 
> 
> > Would you like me to open the "Streamline NonStop Configuration" PR as a defect/enhancement? It's certainly not a nit.
> 
> Enhancement.

Created issue #12858 for this effort. It can be assigned to me.

**Comment by levitte** [comment]
I merged all the commits together into one.  Merged

08073700cc50bcd0df5c0ee68c100e300a320d03 NonStop port updates for 3.0.0.

**Comment by levitte** [comment]
@rsbeckerca, `git commit` has a couple of cool options that make intentions very clear, `--fixup` and `--squash`.  I'd suggest you read about them, they provide clarity of intention with different commits.  It wasn't terribly hard in this case, but the clearer the better, yeah?

**Comment by rsbeckerca** [comment]
> 
> @rsbeckerca, `git commit` has a couple of cool options that make intentions very clear, `--fixup` and `--squash`. I'd suggest you read about them, they provide clarity of intention with different commits. It wasn't terribly hard in this case, but the clearer the better, yeah?

Fun factoid @levitte: I'm actually the platform maintainer for git on NonStop, which is how and why I got into the OpenSSL world. For this PR, the original pull was of a squashed commit of a boat-load of commits rolled into a package. I'm usually prefer to know what's in my sausages, but not for PRs in Open Source contributions. What I missed was how this team modifies pull requests with post-PR commits and then reissues a squash when things get too far down the chain. I'll know for next time, thanks. Git's contribution procedure is at the other end of the spectrum, where each file in a contribution has its own commit (!) - for historical reasons. It's possible that git does not want to depend on a meta-git process to avoid falling into a recursive dependency hole.

**Comment by levitte** [comment]
> Fun factoid @levitte: I'm actually the platform maintainer for git on NonStop

Oh!

**Comment by mspncp** [comment]
> What I missed was how this team modifies pull requests with post-PR commits and then reissues a squash when things get too far down the chain.

O don’t think @levitte meant to say that we do cross-pr-squashes (which we don‘t do).

**Pull request closed without merging**
